themes/default/content/blog/testable-iam-policy-documents/index.md

```diff
@@ -0,0 +1,322 @@
+---
+title: "Testable IAM Policy Documents"
+date: 2021-05-06
+draft: true
+meta_desc:
+meta_image: meta.png
+authors:
+    - thierry-de-pauw
+tags:
+    - AWS
+    - IAM
+---
+
+I was relieved to find Pulumi. Finally, we have testable Infrastructure as Code. We can write fast unit tests that we can execute locally without needing the cloud. However, I was disappointed that Pulumi does not have a proper API for manipulating AWS IAM Policy documents.
+
+<!--more-->
+
+Policy documents are assigned using JSON objects that should follow the AWS
+IAM JSON Policy syntax.
+
+```typescript
+const policy = new aws.iam.Policy("policy", {
+    description: "My test policy",
+    policy: JSON.stringify({
+        Version: "2012-10-17",
+        Statement: [{
+            Action: ["ec2:Describe*"],
+            Effect: "Allow",
+            Resource: "*",
+        }],
+    }),
+});
+```
+
+However, it is perfectly possible to pass an invalid IAM Policy document because there is no validation. You would only notice if it is invalid the minute the policy is applied in the AWS cloud. That is an unreasonably long feedback loop, incurring a significant amount of waiting and time to correction.
```

Comment:
```suggestion
However, it is perfectly possible to pass an invalid IAM Policy document because there is no validation. You would only notice if it is invalid the minute the policy is applied in the AWS cloud. That creates a long feedback loop, incurring a significant amount of waiting and time to correction.
```